### PR TITLE
Use unique_ptr constructor for PluginFactory plugins in RecoTracker

### PR DESCRIPTION
**Pull request by makortel**
Also manage a plugin with `unique_ptr` in `CtfSpecialSeedGenerator`, and use templated parameters for `SeedGeneratorFromRegionHits` constructor (for the migration time) in order to support any mixture of raw pointers and `unique_ptr`'s.

This PR is preparatory work to change the PluginFactory to return a `std::unique_ptr`.

Tested in CMSSW_10_5_X_2019-02-18-1100, no changes expected.

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
@cmsbuild, please test

File: RecoTracker/DebugTools/plugins/CkfDebugTrackCandidateMaker.h [hunk 1]
```diff
@@ -6,6 +6,7 @@
 #include "CkfDebugTrajectoryBuilder.h"
 #include "FWCore/Framework/interface/EDProducer.h"
 #include "DataFormats/TrackReco/interface/SeedStopInfo.h"
+#include "TrackingTools/TrajectoryFiltering/interface/TrajectoryFilter.h"
```
Comment (perrotta) [inline RecoTracker/DebugTools/plugins/CkfDebugTrackCandidateMaker.h]: Why is this needed?

Comment (makortel) [reply RecoTracker/DebugTools/plugins/CkfDebugTrackCandidateMaker.h]: Good question. After looking it more closely, the `CkfDebugTrajectoryBuilder.h` would probably be a more proper placement for the include.

My best guess for why the include addition is needed (gcc fails to compile without one because being unable to generate the destructor of `unique_ptr<TrajectoryFilter>`) is that this PR changes the constructor of `CkfTrajectoryBuilder` to take a `unique_ptr<TrajectoryFilter>` by value
https://github.com/cms-sw/cmssw/pull/25980/files#diff-4d8eac66a767e59ff11481edf3910403
and `CkfDebugTrajectoryBuilder` calls that constructor by passing a `nullptr`
https://github.com/cms-sw/cmssw/blob/fbb6e49f893d3ed1429860e6edb91cd86818e6e9/RecoTracker/DebugTools/plugins/CkfDebugTrajectoryBuilder.h#L11-L12
and (somehow) there the compiler needs to generate the destructor for `unique_ptr`.

I can certainly move the include to `CkfDebugTrajectoryBuilder.h` if you wish.

Comment (perrotta) [reply RecoTracker/DebugTools/plugins/CkfDebugTrackCandidateMaker.h]: Thank you for your explanation.

Yes, after that I think that having it included in ```CkfDebugTrajectoryBuilder.h``` may be more appropriate. Even more appropriate could be, according to me, explicitly defining in the constructor what ```nullptr``` refers to, e.g.
```C++
  CkfDebugTrajectoryBuilder(const edm::ParameterSet& conf):
    CkfTrajectoryBuilder(conf, std::unique_ptr<TrajectoryFilter> filter = nullptr)
```
so that who reads the code can understand why ```TrajectoryFilter.h``` needs to be included.

What do you think?

Comment (perrotta) [reply RecoTracker/DebugTools/plugins/CkfDebugTrackCandidateMaker.h]: Uhm, no: my proposal is not good C++ and it can't compile...
Let stick on your proposal, then, but move the include inside ```CkfDebugTrajectoryBuilder.h``` if that's enough to make it compiling

Comment (makortel) [reply RecoTracker/DebugTools/plugins/CkfDebugTrackCandidateMaker.h]: @perrotta If you want a type to be bound to the `nullptr`, the following works
```cpp
  CkfDebugTrajectoryBuilder(const edm::ParameterSet& conf):
    CkfTrajectoryBuilder(conf, std::unique_ptr<TrajectoryFilter>{})
```
Let me know if you prefer that over the plain `nullptr`.

Comment (perrotta) [reply RecoTracker/DebugTools/plugins/CkfDebugTrackCandidateMaker.h]: Well, my (slight) preference would be to have the type explicitly available, so that nobody has to wonder why the header of ```TrajectoryFilter.h``` was included...
But please proceed as you think it can make better sense.

Comment (makortel) [reply RecoTracker/DebugTools/plugins/CkfDebugTrackCandidateMaker.h]: Done

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by perrotta** [comment]
+1
- Technical
- Jenkins tests pass

**Comment by fabiocos** [comment]
+1
